### PR TITLE
Make EVP_MD_CTX_ctrl() work for legacy use cases.

### PR DESCRIPTION
**Pull request by slontis**
This is still required by engines and digestsign/digestverify

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by beldmit** [comment]
Strongly support.

File: crypto/evp/digest.c [hunk 1]
```diff
@@ -539,30 +539,32 @@ int EVP_MD_CTX_get_params(EVP_MD_CTX *ctx, const OSSL_PARAM params[])
...
+
+            switch (cmd) {
+            case EVP_MD_CTRL_XOF_LEN:
+                if (ctx->digest->set_params == NULL)
+                    break;
+                i = (size_t)p1;
+                params[n++] = OSSL_PARAM_construct_size_t(
+                                  OSSL_DIGEST_PARAM_XOFLEN, &i, &sz);
+                params[n++] = OSSL_PARAM_construct_end();
+                return ctx->digest->set_params(ctx->provctx, params) > 0;
```
Comment (levitte) [inline crypto/evp/digest.c]: Why `> 0`?

File: crypto/evp/digest.c [hunk 2]
```diff
@@ -539,30 +539,32 @@ int EVP_MD_CTX_get_params(EVP_MD_CTX *ctx, const OSSL_PARAM params[])
     return 0;
 }
 
-#if !OPENSSL_API_3
```
Comment (mattcaswell) [inline crypto/evp/digest.c]: I believe some of the documentation says that EVP_MD_CTX_ctrl is deprecated. If we're going down that route then that should be updated.

File: crypto/evp/digest.c (hunk 1 above)
Comment (slontis) [reply crypto/evp/digest.c]: Code has been removed

File: crypto/evp/digest.c (hunk 2 above)
Comment (slontis) [reply crypto/evp/digest.c]: updated

**Comment by slontis** [comment]
ping @mattcaswell

File: crypto/evp/digest.c [hunk 3]
```diff
@@ -539,31 +539,14 @@ int EVP_MD_CTX_get_params(EVP_MD_CTX *ctx, const OSSL_PARAM params[])
...
-            return ctx->digest->set_params(ctx->provctx, params) > 0;
-        case EVP_MD_CTRL_MICALG:
-            if (ctx->digest->get_params == NULL)
-                break;
-            params[n++] = OSSL_PARAM_construct_utf8_string(
-                              OSSL_DIGEST_PARAM_MICALG, p2, p1 ? p1 : 9999,
-                              &sz);
-            params[n++] = OSSL_PARAM_construct_end();
-            return ctx->digest->get_params(ctx->provctx, params);
-        }
```
Comment (mattcaswell) [inline crypto/evp/digest.c]: Why is this deleted? Won't that break upgrading applications that don't know about `EVP_MD_CTX_set_params()`?

File: ssl/s3_enc.c [hunk 4]
```diff
@@ -466,8 +467,12 @@ size_t ssl3_final_finish_mac(SSL *s, const char *sender, size_t len,
...
         OSSL_PARAM digest_cmd_params[3];
+        int cmd;
+        void *msg;
+
+        ssl3_digest_master_key_set_params(s->session, &cmd,
+                                          &msg, digest_cmd_params);
 
-        ssl3_digest_master_key_set_params(s->session, digest_cmd_params);
         if (EVP_DigestUpdate(ctx, sender, len) <= 0
             || EVP_MD_CTX_set_params(ctx, digest_cmd_params) <= 0
```
Comment (levitte) [inline ssl/s3_enc.c]: You're reverting to use `EVP_MD_CTX_ctrl()` in ssl/statem/statem_lib.c...  why not here?

Comment (mattcaswell) [reply ssl/s3_enc.c]: Because in statem_lib.c the operation is a DigestSign/DigestVerify which hasn't been updated to handle providers yet. This is a plain Digest operation which knows how to handle providers.

Comment (levitte) [reply ssl/s3_enc.c]: Ah!

**Comment by asicminer001** [comment]
Duplicate of #

File: ssl/s3_enc.c [hunk 5]
```diff
@@ -412,16 +412,17 @@ int ssl3_digest_cached_records(SSL *s, int keep)
 }
 
 void ssl3_digest_master_key_set_params(const SSL_SESSION *session,
```
Comment (levitte) [inline ssl/s3_enc.c]: Regarding this function, please consider #9126

Comment (slontis) [reply ssl/s3_enc.c]: @levitte  I have merged your changes into this PR (since I needed to test that the ssl3 code actually works).

File: crypto/evp/digest.c (hunk 3 above)
Comment (slontis) [reply crypto/evp/digest.c]: ahh yeah - I was thinking it was only needed for engines and Sign - but I am wrong :).

File: crypto/evp/digest.c [hunk 6]
```diff
@@ -539,30 +539,33 @@ int EVP_MD_CTX_get_params(EVP_MD_CTX *ctx, const OSSL_PARAM params[])
...
+            OSSL_PARAM params[2];
+            size_t i, sz, n = 0;
+
+            switch (cmd) {
+            case EVP_MD_CTRL_XOF_LEN:
+                if (ctx->digest->set_params == NULL)
+                    break;
+                i = (size_t)p1;
+                params[n++] = OSSL_PARAM_construct_size_t(
+                                  OSSL_DIGEST_PARAM_XOFLEN, &i, &sz);
```
Comment (levitte) [inline crypto/evp/digest.c]: I think our coding style says that arguments should be aligned within the enclosing parentheses.  In this case, it might be easier to make a line break after `=`, i.e.:

``` C
                params[n++] =
                    OSSL_PARAM_construct_size_t(OSSL_DIGEST_PARAM_XOFLEN, &i, &sz);

Comment (levitte) [reply crypto/evp/digest.c]: (possibly with another line break among the arguments)

**Comment by levitte** [comment]
Travis shows faults in some of the external project tests...  They should be subject for another PR.

**Comment by slontis** [comment]
> Travis shows faults in some of the external project tests... They should be subject for another PR.

yes this PR is related to getting rid of the SSL3 issues in the extended tests.

File: include/openssl/core_names.h [hunk 7]
```diff
@@ -42,8 +42,7 @@ extern "C" {
 
 /* digest parameters */
 #define OSSL_DIGEST_PARAM_XOFLEN    "xoflen"
-#define OSSL_DIGEST_PARAM_CMD       "cmd"
-#define OSSL_DIGEST_PARAM_MSG       "msg"
+#define OSSL_DIGEST_PARAM_SSL3_MS   "ssl3-ms"
```
Comment (levitte) [inline include/openssl/core_names.h]: Side note:  I made this name up, and am wondering if it's good enough.  It's currently highly internal, but I'm thinking that if anyone else wants to implement something like this, we might want to have something more descriptive.  Do we care much?  Maybe this is also a cue for implementing the SSL/TLS MAC differently, and this might go away at some point.  I actually don't know, but maybe we should make a collective mental note of this?

**Comment by levitte** [comment]
(if anyone wonders how approval works in a case like this, where code from a different PR, #9126 in this case, gets incorporated here, it's really quite simple.  By incorporating #9126, @slontis has implicitly approved it, which is sufficient for that code, and since I'm approving of this PR, that covers the rest of this PR as usual)

**Comment by mattcaswell** [comment]
Travis failure is not relevant.

File: include/openssl/core_names.h (hunk 7 above)
Comment (richsalz) [reply include/openssl/core_names.h]: Its fine, but maybe ssl3ms?  nit.
